org.lflang/src/org/lflang/cli/CliBase.java

```diff
@@ -82,136 +113,144 @@ protected CliBase(String toolName) {
         this.toolName = toolName;
     }
 
-    protected static void cliMain(String toolName, Class<? extends CliBase> toolClass, Io io, String[] args) {
+    protected static void cliMain(
+            String toolName, Class<? extends CliBase> toolClass,
+            Io io, String[] args) {
         // Injector used to obtain Main instance.
         final Injector injector = getInjector(toolName, io);
         // Main instance.
         final CliBase main = injector.getInstance(toolClass);
-        main.runMain(args);
+        // Parse arguments and execute main logic.
+        CommandLine cmd = new CommandLine(main)
+            .setOut(new PrintWriter(io.getOut()))
+            .setErr(new PrintWriter(io.getErr()));
+        int exitCode = cmd.execute(args);
+        io.callSystemExit(exitCode);
     }
 
+    public abstract void run();
```

Comment:
??

Reply:
Piccoli applications are designed to be implemented with the `Runnable`/`Callable` interfaces, and I chose to make our `CliBase` class implement the `Runnable` interface, which is why it needed a `public void run()` method. `Lfc` and `Lff` required different `run()` methods, so I made the one in `CliBase` abstract and wrote custom versions for `Lfc` and `Lff`.

Reply:
Hey, thanks for the patient explanation! I think a short javadoc comment here would be nice (just mentioning that PicoCli requires this so that it doesn't get refactored away). Also an Override annotation is missing.

Also I apologize that my comments are so unpolished, I intended them as notes to myself, but apparently github submitted everything when I answered the other discussion above.

Reply:
No worries, I agree a javadoc there would be helpful. Thanks for your comments!

Reply:
Could you add the javadoc? I think it would be really helpful.

Reply:
Just added it!